python/ray/dag/compiled_dag_node.py

```diff
@@ -471,23 +526,71 @@ def execute(
         if len(kwargs) != 0:
             raise NotImplementedError("Compiled DAGs do not support kwargs")
 
-        input_channel, output_channels = self._get_or_compile()
-        input_channel.write(args[0])
-        return output_channels
+        if self._enable_asyncio:
+            raise ValueError("Use execute_async if enable_asyncio=True")
+
+        self._get_or_compile()
+        self._dag_submitter.write(args[0])
+
+        return self._dag_output_fetcher
+
+    async def execute_async(
+        self,
+        *args,
+        **kwargs,
+    ) -> AwaitableDAGOutput:
+        """Execute this DAG using the compiled execution path.
+
+        NOTE: Not threadsafe.
+
+        Args:
+            args: Args to the InputNode.
+            kwargs: Kwargs to the InputNode. Not supported yet.
+
+        Returns:
+            A list of Channels that can be used to read the DAG result.
+        """
+        # These errors should already be caught during compilation, but just in
+        # case.
+        if len(args) != 1:
+            raise NotImplementedError("Compiled DAGs support exactly one InputNode arg")
+        if len(kwargs) != 0:
+            raise NotImplementedError("Compiled DAGs do not support kwargs")
+
+        if not self._enable_asyncio:
+            raise ValueError("Use execute if enable_asyncio=False")
+
+        self._get_or_compile()
+        async with self._dag_submission_lock:
+            await self._dag_submitter.write(args[0])
+            # Allocate a future that the caller can use to get the result.
```

Comment:
Can you mention the future is set from a background thread of fetcher?

Reply:
Ah actually it is not set from a background thread. Fetching is done in the background thread, and then the result is set in the main thread. But I'll note that down.